### PR TITLE
haxe.iterators.ArrayIterator

### PR DESCRIPTION
**Pull request by RealyUniqueName**
This allows some optimizations to be applied to `Array.iterator` 
E.g. this
```haxe
static function main() {
	var a = [1, 2, 3];
	var r = find(a, i -> i == 2);
	trace(r);
}

public static inline function find<T>(it:Iterable<T>, f:(item:T) -> Bool):Null<T> {
	var result = null;
	for (v in it) {
		if (f(v)) {
			result = v;
			break;
		}
	}
	return result;
}
```
is compiled to this js:
```js
var a = [1,2,3];
var result = null;
var _g_current = 0;
while(_g_current < a.length) {
	var v = a[_g_current++];
	if(v == 2) {
		result = v;
		break;
	}
}
console.log("src/Main.hx:5:",result);
```
notice the absence of `Iterator` allocation.

### CODE REVIEW

**Review by Simn — APPROVED** [review]

File: std/haxe/iterators/ArrayIterator.hx
Comment (nadako) [inline std/haxe/iterators/ArrayIterator.hx:45]: maybe it would be a good time to decide if we want to cache the length or not

Comment (RealyUniqueName) [reply std/haxe/iterators/ArrayIterator.hx:45]: It would be a breaking change.

Comment (Simn) [reply std/haxe/iterators/ArrayIterator.hx:45]: I don't see the point because this is undefined behavior anyway.

Comment (nadako) [reply std/haxe/iterators/ArrayIterator.hx:45]: Well, modification during iteration was always undefined and target-specific. Your change makes it consistent which is good, but is also breaking as it uncaches length for e.g. C# and Java, but this is also fine because, as said - this is officially undefined behaviour. I'd argue that we should cache the len for performance (less field access/getter calls), but maybe it's not worth it, don't know :-/

Comment (RealyUniqueName) [reply std/haxe/iterators/ArrayIterator.hx:45]: Array iteration works without length caching for years (well, except C# and Java when typed as `Iteratable<T>`). It may be unspecified, but it's pretty consistent already even without this PR and that's how it works in existing code bases. Changing that definitely would break someone's projects for no good reason. I think performance gain would be insignificant if any.

**Comment by ncannasse** [comment]
I agree with Alexander

**Comment by RealyUniqueName** [comment]
This change doesn't play well with hashlink because of separate implementations of `Array` at runtime with separate `Iterator` implementations each.
Otherwise making a closure of `Array.iterator` fails in genhl:
```haxe
trace([1,2,3].iterator);
```
```
Main.hx:5: characters 9-19 : Don't know how to cast virtual(hasNext:method:():bool,next:method:():i32) to haxe.iterators.ArrayIterator
```
So I had to fallback to `typedef ArrayIterator<T> = Iterator<T>` for HL.
@ncannasse any ideas how to solve this and use `class ArrayIterator<T>` for all array implementations of HL?

**Comment by ncannasse** [comment]
We don't want to a have single ArrayIterator for HL implementations or else that will create boxing when iterating over basic types (Int/Float/etc.)

**Comment by ncannasse** [comment]
I guess that goes as well for all other native platforms

**Comment by RealyUniqueName** [comment]
That iterator is only applied when array is typed as `Iterable<T>`. I think it erases specific iterator item type anyway.

**Comment by nadako** [comment]
> it erases specific iterator item type anyway.

But `T` *is* the specific item type?

**Comment by RealyUniqueName** [comment]
I mean here `T` is boxed anyway:
```haxe
function some<T>(it:Iterable<T>):T {
  return it.next();
}
```

**Comment by ncannasse** [comment]
Make sure that the fact that ArrayIterator is either a typedef of a class does not break docgen

**Comment by RealyUniqueName** [comment]
I don't like this class/typedef split and I prefer to not merge this if no better solution found.

File: std/hl/types/ArrayObj.hx
Comment (nadako) [inline std/hl/types/ArrayObj.hx]: Why `@:generic`?

Comment (RealyUniqueName) [reply std/hl/types/ArrayObj.hx]: Ah, indeed. Copy-paste leftovers )

**Pull request merged**
